waspc/data/Generator/templates/server/src/auth/utils.ts

```diff
@@ -102,3 +187,64 @@ export async function validateAndGetAdditionalFields(data: {
   }
   return result;
 }
+
+export type EmailProviderData = {
+  password: string;
+  isEmailVerified: boolean;
+  emailVerificationSentAt: Date | null;
+  passwordResetSentAt: Date | null;
+}
+
+export type UsernameProviderData = {
+  password: string;
+}
+
+export type OAuthProviderData = {}
+
+export type ProviderData = {
+  email: EmailProviderData;
+  username: UsernameProviderData;
+  oauth: OAuthProviderData;
+}
+
+export function deserializeProviderData<ProviderName extends keyof ProviderData>(
+  providerData: string,
+  { shouldRemovePasswordField = false }: { shouldRemovePasswordField?: boolean } = {},
+): ProviderData[ProviderName] {
+  // NOTE: We are letting JSON.parse throw an error if the providerData is not valid JSON.
+  let data = JSON.parse(providerData) as ProviderData[ProviderName];
+
+  if (providerDataHasPasswordField(data) && shouldRemovePasswordField) {
+    delete data[PASSWORD_FIELD];
+  }
+
+  return data;
+}
+
+export async function sanitizeAndSerializeProviderData<ProviderName extends keyof ProviderData>(providerData: ProviderData[ProviderName]) {
+  return serializeProviderData(
+    await sanitizeProviderData(providerData)
+  );
+}
+
+async function serializeProviderData<ProviderName extends keyof ProviderData>(providerData: ProviderData[ProviderName]) {
+  return JSON.stringify(providerData);
+}
+
+async function sanitizeProviderData<ProviderName extends keyof ProviderData>(providerData: ProviderData[ProviderName]) {
+  // NOTE: doing a shallow copy here as we expect the providerData to be
```

Comment:
Why do we care about it being deeply copied? It is enough to do shallow copy to be able to modify the password field. So the only reason to do deep copy is because the user of this function expects a completely new object, deeply copied but why would they expect it?

Finally, if it indeed is important, why not just do deep copy right now? It is shallow copy if object is shallow, and it will do what it should do if object is deeper and then you don't need to have this NOTE.

Reply:
You are right, no need to concern ourselves with deeping copying here.

Reply:
Ok, once fixed please write so or throw a :rocket: so I can resolve this one.